### DISCUSSION TITLE
Cog not working?

### DISCUSSION THREAD

**Discussion by JoDotNet** [nextcord/nextcord#704] [Q&A] [answered]
I recently started using cogs, but suddently it just stopped working, when I run the bot theese messages popup:
```
raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
```

Bot runs and everything, but commands won't appear

main.py

```
import nextcord
from nextcord.ext import commands
from nextcord import Interaction


intents = nextcord.Intents.default()
intents.members = True

bot = commands.Bot(intents = intents)


@bot.event
async def on_ready():
    print(f'Logged in as: {bot.user}')


initial_extentions = []

for filename in os.listdir('./cogs'):
    if filename.endswith('.py'):
        initial_extentions.append("cogs." + filename[:-3])

if __name__ == '__main__':
    for extention in initial_extentions:
        bot.load_extension(extention)



bot.run(os.environ['TOKEN'])
```

Administrative.py

```
import nextcord
from nextcord.ext import commands, application_checks
from nextcord import Interaction


class Administrative(commands.Cog):

    def __init__(self, bot: commands.Bot):
        self.bot = bot


    # Ping
    @nextcord.slash_command(guild_ids=[GUILD_ID], name = "ping", description = "Check if the bot responds normally")
    @application_checks.has_any_role(p_Administrator)
    async def ping(self, ctx: Interaction):
        await ctx.send("Pong :ping_pong:", delete_after = 30)


def setup(bot):
    bot.add_cog(Administrative(bot))
```

**Comment by spifory** [comment]
Can you send the code where the error comes from, the traceback would show the line where the error is coming from, just so it's more simple to help you. Also, please note that you can ask for help in the [Discord Server](https://discord.gg/nextcord) so that you do not have to wait for people on GitHub to respond.

**Reply by JoDotNet** [reply]
`79 | bot.run(os.environ["TOKEN"])`

**Comment by DenverCoder1** [comment]
Can you show the full error?

This can sometimes happen if you don't pass a valid token (and it should say so in the full traceback)

This could be caused by not loading environment variables. I'd recommend checking out [python-dotenv](https://pypi.org/project/python-dotenv/)

**Reply by JoDotNet** [reply]
![image](https://user-images.githubusercontent.com/82283920/175617883-2e8d8f3e-94bc-4d4d-9b56-07ff049a9e07.png)

I'm using Heroku

**Reply by spifory** [reply]
> ![image](https://user-images.githubusercontent.com/82283920/175617883-2e8d8f3e-94bc-4d4d-9b56-07ff049a9e07.png)
> 
> I'm using Heroku


Yes, as Denver had said, you would need to use [python-dotenv](https://pypi.org/project/python-dotenv/) in order to load enviroment variables.

**Reply by JoDotNet** [reply]
That's very weird, because I've always done it the same way.

**Reply by DenverCoder1** [reply]
On heroku, there is a Config Vars section in the settings

**Reply by DenverCoder1** [reply]
![image](https://user-images.githubusercontent.com/20955511/175629712-1f6311a0-11da-438c-9f74-aec7c36fede2.png)

**Reply by DenverCoder1** [reply]
You don't need to use python-dotenv, but you need some way of exporting your variables to your bot. Just adding them in the Heroku settings is sufficient for it to run on Heroku with your code.

**Reply by JoDotNet** [reply]
Hey, I tried the python-dotenv method and the orginal method, but none seem to work. The Heroku config is correct.
https://github.com/nextcord/nextcord/blob/master/examples/secure_token_storage.py/#L13
https://www.youtube.com/watch?v=FUTu943XEK8&ab_channel=Dannycademy

**Reply by DenverCoder1** [reply]
What version of nextcord are you using, and also, did you make sure the name of the key matches the one in your code eg. calling it `TOKEN` rather than `DISCORD_TOKEN`?

**Reply by JoDotNet** [reply]
Name is correct, using using this version in requirements: nextcord==2.0.0b4

**Reply by DenverCoder1** [reply]
You can try with nextcord==2.0.0rc2, as that is the latest

**Reply by DenverCoder1** [reply]
Another thing you can check is that you have enabled privileged intents in the Discord developer portal (eg. members intent needs to be on)

**Reply by JoDotNet** [reply]
Updated the nextcord, member intent is already on. Still not working, same error.

**Reply by DenverCoder1** [reply]
What packages do you have in your requirements? Anything besides nextcord?

**Reply by JoDotNet** [reply]
```
nextcord==2.0.0rc2
datetime
pytz
python-dotenv
```

**Reply by DenverCoder1** [reply]
> ![image](https://user-images.githubusercontent.com/82283920/175617883-2e8d8f3e-94bc-4d4d-9b56-07ff049a9e07.png)
> 
> I'm using Heroku

The line above this should probably also have an explanation of the error, are you able to scroll up further?

**Reply by JoDotNet** [reply]
That's the top

**Reply by DenverCoder1** [reply]
There are more logs that it's hiding then. You can access them with the [Heroku CLI](https://devcenter.heroku.com/articles/heroku-cli), or get an add-on for logs such as [Papertrail](https://elements.heroku.com/addons/papertrail)

**Reply by DenverCoder1** [reply]
> ```
> nextcord==2.0.0rc2
> datetime
> pytz
> python-dotenv
> ```

datetime is a built-in module btw, no need to require it. That shouldn't be causing your error though.

**Reply by JoDotNet** [reply]
Using the Papertrail there is no additional information

**Comment by JoDotNet** [comment, marked as answer]
I created a new version with the exact same things, but now it works. Only problem is that _**on_ready**_ it doesn't print my message
https://github.com/JoDotNet/Redon-Bot-Test

**Reply by DenverCoder1** [reply]
There is no on_ready in that repo

on_start is not a Discord event

**Reply by JoDotNet** [reply]
meaning I would need to do what?

**Reply by DenverCoder1** [reply]
Create an on_ready event like you've done before

**Reply by JoDotNet** [reply]
Yeah xD, just realized it.

**Reply by JoDotNet** [reply]
Thank you!
